### DISCUSSION TITLE
What is the reason behind Hungarian notation in the interfaces?

### DISCUSSION THREAD

**Discussion by Orfac** [dotnet/csharplang#4471] [Q&A] [answered]
What is the reason behind Hungarian notation in the interfaces?
Why it is generally right thing to call them in a such way. Why it was added to microsoft guidleines?
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-classes-structs-and-interfaces?WT.mc_id=DOP-MVP-5001077

**Comment by qrli** [comment, marked as answer]
Simply because more people were used to it and prefered it, at that time when .NET was created.
Naming is hard. Java names interfaces by adjectives, like `Enumerable`, to differentiate from classes. But it is not always easy to find a good adjective word. With `I` prefix, it is easy to come up names like `IPlugin`, `IAuthorizationHandler`...

**Comment by jnm2** [comment]
(Type parameters are named this way as well: `TKey`, `TValue`.)

**Comment by JustNrik** [comment]
c# doesn't have a keyword to implement interfaces, so you can confuse them with a class.

```cs
public class ArrayList : List
{
    ...
}
```

Does it inherit a List class or does it implement a List interface? Since you can't really tell, hungarian notation comes and adds I prefix to explicitly state it's an interface.

That's what I think the reason could be though.

**Comment by pinkfloydx33** [comment]
These are basically the same responses recieved yesterday on [SO](https://stackoverflow.com/q/66377166/491907). Are you looking for some sort of different answer?

**Reply by Orfac** [reply]
Yesterday I figured out how it works. Today I was wondering to see what was the reason and what was the consequence. 
And now, to conclude my opinion is

People who created c# we're old and they used to add postfix 'I' for interfaces.
Also they wanted to develop inheritance mechanism so you don't need to explicitly write - you extending class or implementing an interface

So they added a rule to codestyle to use 'I'

**Reply by theunrepentantgeek** [reply]
It's much less "they were old" (which btw is vaguely insulting) and more "they were experienced". 

There are lots of conventions that can seem useless when dealing with toy systems (say, smaller than 25 kloc) that become vital for understanding larger systems (say, larger than 1 Mloc). 

Making interface visually recognisable (by adopting the common prefix `I`) is helpful because interfaces often represent core concepts of the system - recognising and understanding the interfaces can be key to understanding the entire system.

**Reply by Orfac** [reply]
Of course, by too old, I meant they used practises from 20 years ago, not trying to hate them or smth else (sorry for misunderstanding).
If that really helps in large projects, that seems they added 'I' to guidelines just as helpful tip.
Thanks for explanation.
